### PR TITLE
Fix wrong device_id computation in identifier_from_devid

### PR DESCRIPTION
**Pull request by janciesko**
Fixes: https://github.com/kokkos/kokkos/issues/4455

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
How do we know this is less wrong than the code that was there before?

**Comment by janciesko** [comment]
By reading. Read it :)

**Comment by janciesko** [comment]
I'll add a code comment that explains the bit range assignments.

**Comment by janciesko** [comment]
```#include <stdint.h>
#include <stdio.h>

struct ExecutionSpaceIdentifier {
  uint32_t type;
  uint32_t device_id;
  uint32_t instance_id;
};
// Type your code here, or load an example.
ExecutionSpaceIdentifier one(const uint32_t in) {
      return {(in >> 24),  // first 8 bits
          ((in & 0x00FFFFFF) >> 17),  // next 7 bits
           (in & 0x0001FFFF)}; // last 17 bits
}

// Type your code here, or load an example.
ExecutionSpaceIdentifier two(const uint32_t in) {
return {(in >> 24),
          (~((uint32_t(-1)) << 7)) & (in >> 17),
          (~((uint32_t(-1)) << 17)) & in};
}

int main(int c, char ** args)
{
ExecutionSpaceIdentifier a = one(0x12342334);
ExecutionSpaceIdentifier b = two(0x12342334);
printf("%i, %i, %i, %i, %i, %i\n", a.type, a.device_id, a.instance_id, b.type, b.device_id, b.instance_id);
return 1;
}
```

@dalg24 This shows equivalence.

**Comment by janciesko** [comment]
Retest this please.

**Comment by janciesko** [comment]
retest this please

**Comment by janciesko** [comment]
Added unit test coverage

File: core/src/impl/Kokkos_Profiling_Interface.hpp [hunk 1]
```diff
@@ -100,21 +100,19 @@ inline ExecutionSpaceIdentifier identifier_from_devid(const uint32_t in) {
   // out.type = in >> 24;
   // out.device_id = in >> 17;
   // out.instance_id = ((uint32_t(-1)) << 17 ) & in;
```
Comment (cz4rs) [inline core/src/impl/Kokkos_Profiling_Interface.hpp:102]: This comment is outdated and no longer necessary.

File: core/src/impl/Kokkos_Profiling_Interface.hpp [hunk 2]
```diff
@@ -100,21 +100,19 @@ inline ExecutionSpaceIdentifier identifier_from_devid(const uint32_t in) {
...
 struct DeviceTypeTraits;
 
-constexpr const size_t device_type_bits = 8;
-constexpr const size_t instance_bits    = 24;
 template <typename ExecutionSpace>
 constexpr uint32_t device_id_root() {
   constexpr auto device_id =
       static_cast<uint32_t>(DeviceTypeTraits<ExecutionSpace>::id);
-  return (device_id << instance_bits);
+  return (device_id << 17);
```
Comment (cz4rs) [inline core/src/impl/Kokkos_Profiling_Interface.hpp:115]: Magic number: `instance_bits` was easier to read.

**Review by cz4rs — APPROVED** [review]
Looks good to me (only minor nitpicking in the comments).

File: core/unit_test/tools/TestEventCorrectness.hpp [hunk 3]
```diff
@@ -229,6 +229,25 @@ TEST(kokkosp, test_multiple_default_instances) {
   });
 }
 
+/**
+ * Test that device_id() and identifier_from_devid(id) are reciproce
+ * operations
```
Comment (masterleinad) [inline core/unit_test/tools/TestEventCorrectness.hpp:234]:
```suggestion
 * Test that device_id() and identifier_from_devid(id) are reciprocal
 * operations
```

File: core/src/impl/Kokkos_Profiling_Interface.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_Profiling_Interface.hpp:115]: I agree. I think we should reintroduce `device_type_bits` and `instance_bits` and also use them in `identifier_from_devid`.

**Comment by janciesko** [comment]
Hmm, @cz4rs, you'd prefer something like this:

```
constexpr const uint32_t num_type_bits     = 8;
constexpr const uint32_t num_device_bits   = 7;
constexpr const uint32_t num_instance_bits = 17;
constexpr const uint32_t num_avail_bits    = sizeof(uint32_t) * CHAR_BIT;


inline ExecutionSpaceIdentifier identifier_from_devid(const uint32_t in) {

  constexpr const uint32_t shift1 = num_avail_bits - num_type_bits;
  constexpr const uint32_t shift2 = shift1 - num_device_bits;

  return {devicetype_from_uint32t(in >> shift1),                   /*First 8 bits*/
          (~((uint32_t(-1)) << num_device_bits)) & (in >> shift2), /*Next 7 bits */
          (~((uint32_t(-1)) << shift2)) & in};                     /*Last 17 bits*/
}

**Review by masterleinad — APPROVED** [review]

**Review by cz4rs — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

**Comment by khuck** [comment]
@janciesko I am working with the current development branch of Kokkos and the ROCm 5.0.2 compiler... and for my HIP kernels, I am now getting the wrong device type and id.  Instead of getting HIP and device 0, I am getting Serial and device 3.  I have updated my code to match what is in the PR test example, but I'm still getting the wrong info.  Did something else change?

**Comment by khuck** [comment]
btw, I know that Kokkos git hash e5a8acce571b0deeb04b8eb53b7bb2aaa9ebbb87 gives what I am expecting...

**Comment by janciesko** [comment]
@Rombur - do you maybe know what might have changed in the HIP backend in respect to obtaining device Id and device type?

**Comment by Rombur** [comment]
No I don't know anything that could have change that

**Comment by khuck** [comment]
@janciesko @Rombur With the current master, I am running HIP kernels on device 0.  The "device ID" passed to our profiler has the value 393217.  That is being translated (by the above code) as a Serial kernel (0), on device 3, instance 1.  That is incorrect...is the device getting constructed differently by Kokkos, or being interpreted incorrectly?

393217 is:
```
0000 0000 0000 0110 0000 0000 0000 0001
| type  | |device||      instance     |
```
The value should be:
50331649:
```
0000 0011 0000 0000 0000 0000 0000 0001
| type  | |device||      instance     |
```
...right?  This has changed since the last Kokkos release (3.5).

**Pull request merged**